auth/user.go

```diff
@@ -455,18 +458,28 @@ func (user *userImpl) SetPassword(password string) {
 //////// CHANNEL ACCESS:
 
 func (user *userImpl) GetRoles() []Role {
+	user.getRolesLock.RLock()
 	if user.roles == nil {
-		roles := make([]Role, 0, len(user.RoleNames()))
-		for name := range user.RoleNames() {
-			role, err := user.auth.GetRole(name)
-			//base.Infof(base.KeyAccess, "User %s role %q = %v", base.UD(user.Name_), base.UD(name), base.UD(role))
-			if err != nil {
-				panic(fmt.Sprintf("Error getting user role %q: %v", name, err))
-			} else if role != nil {
-				roles = append(roles, role)
+		// Upgrade lock
+		user.getRolesLock.RUnlock()
+		user.getRolesLock.Lock()
+		if user.roles == nil {
+			roles := make([]Role, 0, len(user.RoleNames()))
+			for name := range user.RoleNames() {
+				role, err := user.auth.GetRole(name)
+				//base.Infof(base.KeyAccess, "User %s role %q = %v", base.UD(user.Name_), base.UD(name), base.UD(role))
+				if err != nil {
+					user.getRolesLock.Unlock()
+					panic(fmt.Sprintf("Error getting user role %q: %v", name, err))
+				} else if role != nil {
+					roles = append(roles, role)
+				}
 			}
+			user.roles = roles
 		}
-		user.roles = roles
+		user.getRolesLock.Unlock()
+	} else {
+		user.getRolesLock.RUnlock()
 	}
 	return user.roles
```

Comment:
We're still reading `user.roles` outside of a lock here - so the race isn't fully fixed.

I think you'll have to read it into a variable whilst holding a lock, and return that variable. That would prevent a writer from changing to a new slice on line 478 underneath a read on 484